### PR TITLE
Allow dataloader to accept a custom memory pinning function

### PR DESCRIPTION
**Pull request by mcarilli**
Currently, the `pin_memory_batch` function in the dataloader will return a batch comprised of any unrecognized type without pinning the data, because it doesn't know how.

This behavior was preventing us from overlapping data prefetching in Mask-RCNN, whose custom `collate_fn` returns a custom batch type.  

The present PR adds the ability for the user to pass a `pin_fn` alongside any custom `collate_fn` to handle such custom types.

### CODE REVIEW

**Comment by weiyangfb** [comment]
cc @soumith @fmassa

File: torch/utils/data/dataloader.py
Comment (fmassa) [inline torch/utils/data/dataloader.py]: What about replacing the `pin_memory_batch` by `pin_fn`, and by default have it be `pin_memory_batch`?

File: torch/utils/data/dataloader.py
Comment (fmassa) [inline torch/utils/data/dataloader.py]: this is not necessary if we do `pin_fn` defaulting to `pin_memory_batch`

**Review by fmassa — COMMENTED** [review]
This generally looks good to me, thanks!

I have one minor comment: in the same way as we have the `collate_fn` take by default the `default_collate`, we could also have the `pin_fn` by default take the `pin_memory_batch`.

File: torch/utils/data/dataloader.py
Comment (mcarilli) [reply torch/utils/data/dataloader.py]: Fine by me.

**Review by soumith — APPROVED** [review]

**Review by fmassa — APPROVED** [review]
LGTM, thanks!

But I'd rather have a second pair of eyes checking the diff to see if I haven't missed anything.

File: torch/utils/data/dataloader.py
Comment (apaszke) [inline torch/utils/data/dataloader.py]: Unfortunately adding an argument in the middle of the signature is not a backward compatible change 😕

**Review by apaszke — APPROVED** [review]
LGTM, but we might want to move the new argument to the end of the list

File: torch/utils/data/dataloader.py
Comment (soumith) [reply torch/utils/data/dataloader.py]: i'm not sure we need this change. the data loader has so many arguments, no one uses it as *args, only as kwargs

**Comment by yf225** [comment]
@mcarilli This seems to cause intermittent CUDA OOM error on Windows (example: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-test2/16171//consoleFull). Do you mind trying to run this patch on Windows and see where the issue is? Thanks!

**Comment by mcarilli** [comment]
@yf225 Memory pinning is a purely CPU-side operation, it shouldn't cause any OOM errors on the device side.  Also, the tests I included (`test_custom_pin_fn` and `test_custom_pin_fn_worker`) operate purely on CPU-side data, and they seem to be fine in the CI log you sent.  What makes you think this PR is at fault?

**Comment by yf225** [comment]
@mcarilli There are a couple indicators:
1. This is the latest PR that touches the dataloader code (the previous one is 10 days ago which doesn't match the time when the CUDA OOM error first happened, which is roughly 3 days ago after this PR is merged, based on https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master)
2. In one of the previous CI runs for this PR there is a CUDA OOM error: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-trigger/30180/ which looks the same as the errors on master.
3. In https://github.com/pytorch/pytorch/pull/14382 I manually rerun the Windows tests multiple times without this commit and the error doesn't seem to show. 

I agreed that memory pinning should not cause OOM errors, but I feel that it would be a good idea to first confirm that master is green after the revert (by checking https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master), and then we figure out why it happens.

**Comment by mcarilli** [comment]
@yf225 Ok, I guess I will wait to see if master becomes green.  I don‘t have the resources to do comprehensive windows testing on my end and I‘m still mystified why the pinning PR should cause cuda OOMs, especially since my tests are the only two places the non-default pinning logic is used at all.  Please keep me updated...

**Pull request closed without merging**
